### PR TITLE
Jormun: add support for async call to kraken

### PR DESCRIPTION
**Pull request by kinnou02**
There is one major warning: each green thread has it own thread local
storage so stuff like flask.request or flask.g won't work as expected.

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 1]
```diff
@@ -713,13 +715,20 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
-            self.nb_kraken_calls += 1
 
-            local_resp = instance.send_and_receive(pb_request)
+            #we spawn a new green thread, it won't have access to our thread local request object so we set request_id
+            futures[(dep_mode, arr_mode)] = gevent.spawn(instance.send_and_receive,
+                                                         pb_request,
+                                                         request_id=flask.request.id)
 
+        gevent.joinall(futures.values())
+        for (dep_mode, arr_mode), future in futures.items():
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: Without the log line 727|736 I can improve thing a little bit by not waiting that all threads had finished before starting to handle the results.
But i like this log :(

Comment (TeXitoi) [reply source/jormungandr/jormungandr/scenarios/new_default.py]: I don't understand why you can't log if you don't joinall.

Comment (kinnou02) [reply source/jormungandr/jormungandr/scenarios/new_default.py]: I can use `iwait()` for working on them as they finish, this function take a list of future/greenlet.
If I use it I won't be able to know what are the mode used as fallback for this call

Comment (TeXitoi) [reply source/jormungandr/jormungandr/scenarios/new_default.py]: You can just remove the joinall. `future.get()` will wait. That's not as optimal as `iwait()`, but better than `joinall()`.

Or you can spawn a closure returning `(dep_mode, arr_mode, instance.send_and_receive(...))`

**Comment by TeXitoi** [comment]
OK for me.

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 2]
```diff
@@ -713,21 +715,27 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+        def worker(dep_mode, arr_mode, instance, request, request_id):
+            return (dep_mode, arr_mode, instance.send_and_receive(request, request_id=request_id))
+
         for dep_mode, arr_mode in krakens_call:
             pb_request = create_pb_request(request_type, request, dep_mode, arr_mode)
-            self.nb_kraken_calls += 1
 
-            local_resp = instance.send_and_receive(pb_request)
+            #we spawn a new green thread, it won't have access to our thread local request object so we set request_id
+            futures.append(gevent.spawn(worker, dep_mode, arr_mode, instance, pb_request, request_id=flask.request.id))
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: why don't you do a deepcopy in this loop?

**Comment by xlqian** [comment]
:+1:

**Pull request merged**